Sources/NIO/Codec.swift

```diff
@@ -674,3 +674,118 @@ extension MessageToByteHandler {
         }
     }
 }
+
+enum ByteToMessageDecoderVerificationError<OutputType>: Error {
+    case wrongDecodingResult(actual: OutputType, expected: OutputType)
+    case unexpectedDecodingResult(OutputType)
+    case leftOversOnDeconstructingChannel(inbound: [NIOAny], outbound: [NIOAny], pendingOutbound: [NIOAny])
+}
+
+extension ByteToMessageHandler {
+    /// Verifies a `ByteToMessageDecoder` by performing a number of tests.
+    ///
+    /// This method is mostly useful in unit tests for `ByteToMessageDecoder`s. It feeds the inputs from
+    /// `inputOutputPairs` into the decoder in various ways and expects the decoder to produce the outputs from
+    /// `inputOutputPairs`.
+    ///
+    /// The verification is performing various tests, for example:
+    ///
+    ///  - drip feeding the bytes, one by one
+    ///  - sending many messages in one `ByteBuffer`
+    ///  - sending each complete message in one `ByteBuffer`
+    ///
+    /// For `ExampleDecoder` that produces `ExampleDecoderOutput`s you would use this method the following way:
+    ///
+    ///     var exampleInput1 = channel.allocator.buffer(capacity: 16)
+    ///     exampleInput1.writeString("example-in1")
+    ///     var exampleInput2 = channel.allocator.buffer(capacity: 16)
+    ///     exampleInput2.writeString("example-in2")
+    ///     let expectedInOuts = [(exampleInput1, ExampleDecoderOutput("1")),
+    ///                           (exampleInput2, ExampleDecoderOutput("2"))
+    ///                          ]
+    ///     XCTAssertNoThrow(try ByteToMessageHandler<ExampleDecoder>.verifyDecoder(inputOutputPairs: expectedInOuts,
+    ///                                                                             decoderFactory: { ExampleDecoder() }))
+    public static func verifyDecoder<Decoder: ByteToMessageDecoder>(inputOutputPairs: [(ByteBuffer, [Decoder.InboundOut])],
+                                                                    decoderFactory: @escaping () -> Decoder) throws where Decoder.InboundOut: Equatable {
+        typealias Out = Decoder.InboundOut
+
+        func verifySimple(channel: EmbeddedChannel) throws {
+            for (input, expectedOutputs) in inputOutputPairs.shuffled() {
```

Comment:
Wouldn't this make tests non-deterministic?

Reply:
@ianpartridge Yes, it does. The expectation is that B2MDs shouldn't mind that and if they do that's a bug. I could add (in the error) the specific input that was sent maybe?

It's a similar issue like when relying on the order of the key/value pairs in a dictionary I guess.

Reply:
Yes I'm just thinking about how users could debug their B2MD when faced with an order-dependent test failure.

Reply:
@ianpartridge well, it's one B2MD that they'd need to look at and it needs to be state that persists across two messages (of which there should be very little). However, the whole thing could be solved by just giving the user the data in a failed attempt. Then they can make that  particular input a regression test by itself.

Reply:
Yep, sounds good to me.

Reply:
Can we instead accept a hash to affect the shuffling, and then print that hash in the event of test failure?

Reply:
@Lukasa Sure, that works too.

Reply:
@Lukasa / @ianpartridge I added the delivery of the inputs, does that work?